### ISSUE TITLE
variants matrix is not built.

### ISSUE THREAD

**Issue by rainwoodman** [conda/conda-build#2593]
I declared a variant matrix as documented here:

https://conda.io/docs/user-guide/tasks/build-packages/variants.html#general-pinning-examples

But when I run

conda build -m variants/python-3.6.yaml platform/mpi4py

Only one variant for mpich2 is built. I tried this on 3.0.31, 3.1.0, and 3.1.3. 

Here is the recipe:

[troubled-mpi4py.tar.gz](https://github.com/conda/conda-build/files/1569925/troubled-mpi4py.tar.gz)

**Comment by rainwoodman** [comment]
I had a typo `feature` should be `features` but it didn't affect reproducing this bug.

**Comment by rainwoodman** [comment]
It appears that if I do not supply mpich2 in the list, I do get multiple variants:
```
mpi:
  - mpich3
  - mpich1

conda build --output -m variants/mpich/python-3.6.yaml  platform/mpi4py
Returning non-final recipe for mpi4py-3.0.0-py36_mpich3_4; one or more dependencies was unsatisfiable:
Build: mpich3
Host: []
WARNING:conda_build.render:Returning non-final recipe for mpi4py-3.0.0-py36_mpich3_4; one or more dependencies was unsatisfiable:
Build: mpich3
Host: []
Returning non-final recipe for mpi4py-3.0.0-py36_mpich1_4; one or more dependencies was unsatisfiable:
Build: mpich1
Host: []
WARNING:conda_build.render:Returning non-final recipe for mpi4py-3.0.0-py36_mpich1_4; one or more dependencies was unsatisfiable:
Build: mpich1
Host: []
/home/yfeng1/anaconda3/install/conda-bld/linux-64/mpi4py-3.0.0-py36_mpich1h8e6f190_4.tar.bz2
/home/yfeng1/anaconda3/install/conda-bld/linux-64/mpi4py-3.0.0-py36_mpich3hace8075_4.tar.bz2

```

But if I supply mpich2 in the list, I only get mpich2.
```
mpi:
  - mpich2
  - mpich1

conda build --output -m variants/mpich/python-3.6.yaml  platform/mpi4py
/home/yfeng1/anaconda3/install/conda-bld/linux-64/mpi4py-3.0.0-py36_mpich2h49e5514_4.tar.bz2
```

If I do not name the key mpi, I also get multiple entries:
mympi:
  - mpich2
  - mpich1

conda build --output -m variants/mpich/python-3.6.yaml  platform/mpi4py
/home/yfeng1/anaconda3/install/conda-bld/linux-64/mpi4py-3.0.0-py36_mpich2h8eecca6_4.tar.bz2
/home/yfeng1/anaconda3/install/conda-bld/linux-64/mpi4py-3.0.0-py36_mpichhe508baa_4.tar.bz2
```

It appears that somehow conda-build has a built-in mpi variable which is set to mpich2; and overriding it breaks variants.

**Comment by msarahan** [comment]
there's nothing special about the name mpi that I know of.  There is a bug here, though, because your example should work.  Thanks for reporting.  I will look into it.

**Comment by rainwoodman** [comment]
Thanks for looking into the bug. 

The things is I couldn't find the mpi override anywhere neither -- my system is relatively clean..

Could you remind me if the list of file names that conda-build parses listed in the doc is exhaustive? (here is the link)

https://conda.io/docs/user-guide/tasks/build-packages/variants.html#creating-conda-build-variant-config-files

**Comment by rainwoodman** [comment]
I found it. I have a file 'conda_build_config.yaml' that declared mpi as
```
mpi:
  -- mpich2
```

in the current directory.

**Comment by msarahan** [comment]
I believe this was a bug that #2595 fixed.  Rendering your recipe from your zip file, I now get:

```
CONDA_SUBDIR=linux-64 conda render platform/mpi4py -m variants/python-3.6.yaml
setting build platform. This is only useful when pretending to be on another another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
Returning non-final recipe for mpi4py-3.0.0-py36_4; one or more dependencies was unsatisfiable:
Build: mpich
Host: []
package:
    name: mpi4py
    version: 3.0.0
source:
    fn: mpi4py-3.0.0.tar.gz
    md5: bfe19f20cef5e92f6e49e50fb627ee70
    url: https://pypi.python.org/packages/31/27/1288918ac230cc9abc0da17d84d66f3db477757d90b3d6b070d709391a15/mpi4py-3.0.0.tar.gz
build:
    detect_binary_files_with_prefix: 'true'
    number: '4'
requirements:
    build:
        - binutils_impl_linux-64 2.28.1 h04c84fa_2
        - ca-certificates 2017.08.26 h1d4fec5_0
        - libgcc-ng 7.2.0 h7cc24e2_2
        - libstdcxx-ng 7.2.0 h7a57d05_2
        - mpich2 1.4.1p1 0
        - binutils_linux-64 7.2.0 hc67e822_15
        - gcc_impl_linux-64 7.2.0 hc5ce805_2
        - libffi 3.2.1 hd88cf55_4
        - ncurses 6.0 h9df7e31_2
        - openssl 1.0.2n hb7f436b_0
        - tk 8.6.7 hc745277_3
        - xz 5.2.3 h55aa19d_2
        - zlib 1.2.11 ha838bed_2
        - gcc_linux-64 7.2.0 haf1f6fa_15
        - libedit 3.1 heed3624_0
        - readline 7.0 ha6073c6_4
        - sqlite 3.20.1 hb898158_2
        - python 3.6.3 h6c0c0dc_5
        - certifi 2017.11.5 py36hf29ccca_0
        - setuptools 36.5.0 py36he42e2e1_0
        - wheel 0.30.0 py36hfd4bba0_1
        - pip 9.0.1 py36h6c6f9ce_4
    run:
        - mpich2 >=1.4.1p.1,<1.5.0a0
        - python >=3.6,<3.7.0a0
test:
    imports:
        - mpi4py
        - mpi4py.MPI
about:
    description: 'MPI for Python provides bindings of the Message Passing Interface
        (MPI)

        standard for the Python programming language, allowing any Python program

        to exploit multiple processors.

        '
    dev_url: https://bitbucket.org/mpi4py/mpi4py
    doc_url: http://mpi4py.readthedocs.org/
    home: http://mpi4py.scipy.org/
    license: BSD
    summary: Provides bindings of the MPI standard for Python
extra:
    copy_test_source_files: true
    final: true

package:
    name: mpi4py
    version: 3.0.0
source:
    fn: mpi4py-3.0.0.tar.gz
    md5: bfe19f20cef5e92f6e49e50fb627ee70
    url: https://pypi.python.org/packages/31/27/1288918ac230cc9abc0da17d84d66f3db477757d90b3d6b070d709391a15/mpi4py-3.0.0.tar.gz
build:
    detect_binary_files_with_prefix: 'true'
    number: '4'
requirements:
    build:
        - python 3.6.*
        - gcc_linux-64
        - mpich
    run:
        - python >=3.6,<3.7.0a0
        - mpich
test:
    imports:
        - mpi4py
        - mpi4py.MPI
about:
    description: 'MPI for Python provides bindings of the Message Passing Interface
        (MPI)

        standard for the Python programming language, allowing any Python program

        to exploit multiple processors.

        '
    dev_url: https://bitbucket.org/mpi4py/mpi4py
    doc_url: http://mpi4py.readthedocs.org/
    home: http://mpi4py.scipy.org/
    license: BSD
    summary: Provides bindings of the MPI standard for Python
extra:
    copy_test_source_files: true
    final: true

 msarahan@0109-msarahan  ~/Downloads/troubled-mpi4py  CONDA_SUBDIR=linux-64 conda render platform/mpi4py -m variants/python-3.6.yaml --output
setting build platform. This is only useful when pretending to be on another another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
Returning non-final recipe for mpi4py-3.0.0-py36_4; one or more dependencies was unsatisfiable:
Build: mpich
Host: []
/Users/msarahan/miniconda3/conda-bld/linux-64/mpi4py-3.0.0-py36h49e5514_4.tar.bz2
/Users/msarahan/miniconda3/conda-bld/linux-64/mpi4py-3.0.0-py36hf046da1_4.tar.bz2
```

**Comment by msarahan** [comment]
This is fixed in the 3.2.0 release, which is being built and should be available in a few minutes.
